docs/dev/RELEASE.md

```diff
@@ -2,80 +2,64 @@
-   the release, if this is not a pre-release.
-
-7. Push the "Publish" button. This will trigger yet another workflow
-   that will download the published artifacts (from both GitHub and NuGet)
-   and run a smoke test on multiple platforms. 
-   Again you can watch for this workflow at
-   https://github.com/dafny-lang/dafny/actions.
-
-8. If preparing a pre-release, stop here, as
-   the following steps declare the release as the latest version, which
-   is not the intention.
-
-9. If something goes wrong, delete the tag and release in GitHub, fix the
-   problem and try again.
-
-10. Update the Homebrew formula for Dafny (see below).
-    Note that it is fine to leave this for the next day,
-    and other members of the community may update the formula
-    in the meantime anyway.
-
-11. Announce the new release to the world.
+0. Ensure that you are in a clean and up-to-date repository, with the `master`
+   branch checked out and up-to-date and no uncommitted changes.
+
+1. Select a version number `$VER` (e.g., "3.0.0" or "3.0.0-alpha"), then run
+   `Scripts/release.py $VER prepare` from the root of the repository.  The
+   script will check that the repository is in a good state, create and check
+   out a new release branch, update `Source/version.cs` and `RELEASE_NOTES.md`,
+   prepare a release commit, and push it.
+
+2. Kick off the deep test suite by navigating to
+   <https://github.com/dafny-lang/dafny/actions/workflows/deep-tests.yml>,
+   clicking the "Run workflow" dropdown, selecting the newly created branch, and
+   clicking the "Run workflow" button. The automation for releasing below will
+   check for a run of this workflow on the exact commit to release.  (TODO:
+   Run this automatically as part of the prepare-release script.)
+
+3. Once the the tests complete, run `Scripts/release.py $VER release` from the
+   root of the repository.  The script will tag the current commit and push
+   it. (TODO: Merge with the two steps above.)  A GitHub action will
+   automatically run in reaction to the tag being pushed, which will build the
+   artifacts and reference manual and then create a draft GitHub release. You
+   can find and watch the progress of this workflow at
+   <https://github.com/dafny-lang/dafny/actions>.
+
+4. Once the action completes, you should find the draft release at
+   <https://github.com/dafny-lang/dafny/releases>. Edit the release body to add in
+   the release notes from `RELEASE_NOTES.md`.  If this is not a pre-release,
+   check the box to create a new discussion based on the release.
+
+5. Push the "Publish" button. This will trigger yet another workflow
+   that will download the published artifacts and run a smoke test
+   on multiple platforms. Again you can watch for this workflow at
+   <https://github.com/dafny-lang/dafny/actions>.
+
+6. Create a pull request to merge the newly created branch into `master` (the
+   script will give you a link to do so).  Get it approved and merged.
```

Comment:
I don't love the idea of releasing from a branch before merging it to mainline. Could we flip the order of steps to do this first, even if it means customizing the GHA that looks for the deep tests? Are we just trying to save the extra delay on waiting for the CI?

Reply:
Releasing from master forces a freeze to avoid concurrent commits; we've been bitten by that before.  What issue do you see with releasing from a branch?  Is it the security aspect? (I think we already allow releasing by uploading a custom-packaged zip, so I haven't worried about that part)

Reply:
I think it's not only pretty common to release from a branch before merging it (that way, it's even possible to cherry-pick bug fixes independently without introducing new features), but in our case, as Clement said, we already encountered the case when we wanted to release after merging with master; however, there was one more commit on master since we started to prepare the release, and this commit was merged unintentionally with the release.
So although I had wished that releases are made from the history of master branches, in practice, this is dangerous.

Reply:
Okay that's a better reason, I'm happy with this.